pkg/telemetry/telemetry.go

```diff
@@ -0,0 +1,84 @@
+package telemetry
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/rs/zerolog"
+	logger "github.com/rs/zerolog/log"
+)
+
+var (
+	c           *collector
+	metricStore MetricStore
+	log         zerolog.Logger
+
+	mu   = &sync.Mutex{}
+	once sync.Once
+)
+
+func init() {
+	log = logger.With().
+		Str("component", "telemetry").
+		Logger()
+}
+
+// MetricStore specifies the methods for persisting a metric.
+type MetricStore interface {
+	StoreMetric(context.Context, Metric) error
+	Close() error
+}
+
+// SetMetricStore sets the store implementation.
+// Can only be called once, and should be called before Collect.
+func SetMetricStore(s MetricStore) {
+	once.Do(func() {
+		metricStore = s
+	})
+}
+
+// Collect collects the metric by persisting locally for later publication.
+// If Collect is called before setting the metric store, it will simply log the metric without persisting it.
+func Collect(ctx context.Context, metric interface{}) error {
+	mu.Lock()
+	defer mu.Unlock()
+	if metricStore == nil {
```

Comment:
I opted to not return an error with clients who forgot to call `SetMetricStore`. That means that client code will not get interrupted because of a metric that has no store set. The only thing is that the metric will not be persisted. That is good, so I don't have to call `SetMetricStore` for all existing tests.